client/session.go

```diff
@@ -563,3 +595,943 @@ func (s *session) Close() error {
+				reattemptBlocksMetadata[j] = &blocksMetadata{
+					peer: blocks[i].reattempt.peersMetadata[j].peer,
+					id:   blocks[i].reattempt.id,
+					blocks: []blockMetadata{blockMetadata{
+						start:     blocks[i].reattempt.peersMetadata[j].start,
+						size:      blocks[i].reattempt.peersMetadata[j].size,
+						reattempt: blocks[i].reattempt,
+					}},
+				}
+			}
+			// Re-enqueue the block to be fetched
+			enqueue(reattemptBlocksMetadata)
+		}
+	})
+}
+
+type blocksResult struct {
+	sync.RWMutex
+	opts        Options
+	blockOpts   block.Options
+	encoderPool encoding.EncoderPool
+	bytesPool   pool.BytesPool
+	result      bootstrap.ShardResult
+}
+
+func newBlocksResult(opts Options, bootstrapOpts bootstrap.Options) *blocksResult {
+	blockOpts := bootstrapOpts.GetDatabaseBlockOptions()
+	return &blocksResult{
+		opts:        opts,
+		blockOpts:   blockOpts,
+		encoderPool: blockOpts.GetEncoderPool(),
+		bytesPool:   blockOpts.GetBytesPool(),
+		result:      bootstrap.NewShardResult(bootstrapOpts),
+	}
+}
+
+func (r *blocksResult) addBlockFromPeer(id string, block *rpc.Block) error {
+	var (
+		start    = time.Unix(0, block.Start)
+		segments = block.Segments
+		result   = r.blockOpts.GetDatabaseBlockPool().Get()
+	)
+
+	if segments.Merged != nil {
+		size := len(segments.Merged.Head) + len(segments.Merged.Tail)
+		data := r.bytesPool.Get(size)[:size]
+		n := copy(data, segments.Merged.Head)
+		copy(data[n:], segments.Merged.Tail)
+		encoder := r.encoderPool.Get()
+		// TODO(r): Unseal data
+		encoder.ResetSetData(start, data, false)
+		result.Reset(start, encoder)
+
+	} else if segments.Unmerged != nil {
+		// Must merge to provide a single block
+		readers := make([]io.Reader, len(segments.Unmerged))
+		for i := range segments.Unmerged {
+			readers[i] = xio.NewSegmentReader(ts.Segment{
+				Head: segments.Unmerged[i].Head,
+				Tail: segments.Unmerged[i].Tail,
+			})
+		}
+
+		alloc := r.opts.GetReaderIteratorAllocate()
+		iter := encoding.NewMultiReaderIterator(alloc, nil)
+		iter.Reset(readers)
+
+		encoder := r.encoderPool.Get()
+		defer encoder.Close()
+
+		for iter.Next() {
+			dp, unit, annotation := iter.Current()
+			encoder.Encode(dp, unit, annotation)
+		}
+		if err := iter.Err(); err != nil {
+			return err
+		}
+
+		result.Reset(start, encoder)
+
+	} else {
+		return errSessionBadBlockResultFromPeer
+
+	}
+
+	r.Lock()
+	r.result.AddBlock(id, result)
+	r.Unlock()
+
+	return nil
+}
+
+type enqueueChannel struct {
+	enqueued        uint64
+	processed       uint64
+	peersMetadataCh chan []*blocksMetadata
+}
+
+func newEnqueueChannel() *enqueueChannel {
+	return &enqueueChannel{
+		peersMetadataCh: make(chan []*blocksMetadata, 4096),
+	}
+}
+
+func (c *enqueueChannel) enqueue(peersMetadata []*blocksMetadata) {
+	atomic.AddUint64(&c.enqueued, 1)
+	c.peersMetadataCh <- peersMetadata
+}
+
+func (c *enqueueChannel) enqueueDelayed() func([]*blocksMetadata) {
+	atomic.AddUint64(&c.enqueued, 1)
```

Comment:
This is a tricky one to explain.  Essentially before the call to `streamBlocksBatchFromPeer` finishes (which will trigger a `enqueueCh.trackProcessed(1)`) we need to ensure we have tracked another outstanding enqueued piece of work for the retry or else the loop that's checking for no more work could potentially think there's no more work left and close the enqueue channel.